pkg/util/tls/tls.go

```diff
@@ -26,21 +28,29 @@ var (
 
 // RegisterFlagsWithPrefix registers flags with prefix.
 func (cfg *ClientConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, prefix+".tls-enabled", false, "Enable TLS in the client. This flag is implicitly enabled when any other TLS flag is set. This flag will only be respected by the GRPC client, as for the HTTP client the scheme used in the URL will take precedence.")
```

Comment:
> This flag is implicitly enabled when any other TLS flag is set.

I assume this is for backward-compatibility, but it seems weird that using `-client.tls-enabled=false` will not be respected, if used with other tls flags.

We *could* make this a string variable and differentiate between empty string, "true" or "false" (or "enabled"/"disabled"). I don't like this idea, but don't have a better one.

Reply:
> I assume this is for backward-compatibility, but it seems weird that using -client.tls-enabled=false will not be respected, if used with other tls flags.

I haven't considered it, but you're right.

> We could make this a string variable and differentiate between empty string, "true" or "false" (or "enabled"/"disabled"). I don't like this idea, but don't have a better one.

But then in the config file reference documentation the default would be an empty string, which is weird, while having it documented as "false" would be better.

Reply:
> > We could make this a string variable and differentiate between empty string, "true" or "false" (or "enabled"/"disabled"). I don't like this idea, but don't have a better one.
> 
> But then in the config file reference documentation the default would be an empty string, which is weird, while having it documented as "false" would be better.

Yes, it's bad solution. Personally I think we should require explicit use of `tls-enabled` flag, even though it's a breaking change.

Reply:
> Yes, it's bad solution. Personally I think we should require explicit use of tls-enabled flag, even though it's a breaking change.

But one that's discovered and fixed quickly.

Reply:
I've raised a [public question in the CNCF Slack](https://cloud-native.slack.com/archives/CCYDASBLP/p1600339076337200).

Reply:
According to the current status of the Slack discussion, the general consensus is:
1. OK to introduce a breaking change (as Peter suggested)
2. Explicitly mention in `docs/configuration/v1-guarantees.md` that TLS support is experimental (there's a list of experimental features in the `Experimental features` section)

@simonswine WDYT?

Reply:
> 1. OK to introduce a breaking change (as Peter suggested)

While the string based approach is a bit hacky, my main concern is more that a `tls-enable=false` some scenarios will  be quite counter intuitive:

- I am specifying certs+keys, but no `tls-enable=true` => cortex exits with 1. This is not how other TLS enabled software works

- We have someone configuring `tls-enable=false`. But still the HTTP client is not really there to make the call. It is the request that will determine through the URL scheme if it is either `http://` or `https://`.

> 2. Explicitly mention in `docs/configuration/v1-guarantees.md` that TLS support is experimental (there's a list of experimental features in the `Experimental features` section)

This sounds reasonable, happy to add that

Reply:
I think that:

- we should remove "tls_enabled" flag from `tls.ClientConfig` struct.

- for grpc clients, we should put "tls_enabled" bool flag into `grpcclient.Config`, and merge `grpcclient.ConfigWithTLS` with `grpcclient.Config`. When tls_enabled is false, TLS is not configured for grpc connection at all. We have some places in the code where we use `grpcclient.Config` (non-TLS version), which should be fixed, or TLS config put into `grpcclient.Config` directly.

- for HTTP clients (seems to be only ruler in Cortex), they will use TLS based on the scheme, with TLS settings coming from `tls.ClientConfig`.

- etcd client which has its own config should embed `tls.ClientConfig` instead, but keep `tls_enabled` flag.

Reply:
> * I am specifying certs+keys, but no `tls-enable=true` => cortex exits with 1. This is not how other TLS enabled software works

TLS settings should only be consulted when needed (if enabled for grpc/etcd, or always for HTTP clients).

Reply:
> > 1. Explicitly mention in `docs/configuration/v1-guarantees.md` that TLS support is experimental (there's a list of experimental features in the `Experimental features` section)
> 
> This sounds reasonable, happy to add that

Will send a separate PR to make it into Cortex 1.4.

Reply:
@pstibrany: I am currently looking into implementing your suggestions:

> * we should remove "tls_enabled" flag from `tls.ClientConfig` struct.
> * for grpc clients, we should put "tls_enabled" bool flag into `grpcclient.Config`, and merge `grpcclient.ConfigWithTLS` with `grpcclient.Config`. When tls_enabled is false, TLS is not configured for grpc connection at all. We have some places in the code where we use `grpcclient.Config` (non-TLS version), which should be fixed, or TLS config put into `grpcclient.Config` directly.

That's where I am currently at and this has come up so far:

- We have at least a GRPC client that has TLS currently enabled by default as part of the big table support:

https://github.com/cortexproject/cortex/blob/067412e4e79dd1abfe4fbc2ec74e3a13407c3b9d/pkg/chunk/gcp/bigtable_index_client.go#L38

- We seem to use sometimes `tls.ClientConfig` that is used later to generate GRPCConfigs from them e.g.:

https://github.com/cortexproject/cortex/blob/067412e4e79dd1abfe4fbc2ec74e3a13407c3b9d/pkg/ruler/ruler.go#L49

I just wanted to make sure you want me to go forward and for example:
- figure out a way so that we can have the GCP Bigtable with TLS enabled by default
- Add a `TLSEnabled bool` flag to every config part that contains a `tls.ClientConfig`

Reply:
> I just wanted to make sure you want me to go forward and for example:

I still think this is a way to go. 

> 
> * figure out a way so that we can have the GCP Bigtable with TLS enabled by default

This can be achieved by overriding the config default in after this line: https://github.com/cortexproject/cortex/blob/067412e4e79dd1abfe4fbc2ec74e3a13407c3b9d/pkg/chunk/gcp/bigtable_index_client.go#L54

> * Add a `TLSEnabled bool` flag to every config part that contains a `tls.ClientConfig`

I'd say we should replace this with `grpcclient.ConfigWithTLS` instead. WDYT?

Thanks a lot for taking care of this issue!

Reply:
> This can be achieved by overriding the config default after this line:
> https://github.com/cortexproject/cortex/blob/067412e4e79dd1abfe4fbc2ec74e3a13407c3b9d/pkg/chunk/gcp/bigtable_index_client.go#L54
> 

Sorry, I misremembered how we did it last time :) 

Take a look at this method: https://github.com/cortexproject/cortex/blob/85f93ff4abd7af81a5eeeae0e07379dd6ff2071d/pkg/chunk/table_provisioning.go#L33

and how it uses `cfg.ProvisionedWriteThroughput` as default value, which is then set by caller:

https://github.com/cortexproject/cortex/blob/85f93ff4abd7af81a5eeeae0e07379dd6ff2071d/pkg/chunk/purger/table_provisioning.go#L18

Reply:
(I think the same technique should work here, thanks to default value of TLS enabled flag being `false)

Reply:
That sounds all good :+1: 

@pstibrany thanks for the "flags default override example", will keep you posted